model.py

```diff
@@ -0,0 +1,127 @@
+        self.utils.load_validation_set()
+
+        self.net.to(device)
+        self.net.train()
+
+
+        startloop = time.time()
+        lowest_loss = float('inf')
+        for epoch in range(epochs):
+            startepoch = time.time()
+            for i, data in enumerate(train_loader):
+                # Training
+                x_train, y_train = data[0].cuda(device).squeeze(), data[1].cuda(device).squeeze()
+                y_pred = self.net(x_train.float()).squeeze()
+                loss = mse(y_train.float(), y_pred)
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+            if(loss.detach().item()<lowest_loss):
+                lowest_loss = loss.detach().item()
+                print("Saving ->")
+                torch.save({'epoch': epoch, 'model_state_dict': self.net.state_dict(), 'optimizer_state_dict': optimizer.state_dict(), 'scheduler_state_dict': scheduler.state_dict(),'loss': loss}, self.manager.args["save_lowest_loss_model_at"])
+            # Validation
+            self.net.eval()
+            with torch.no_grad():
+                val_y_pred = self.net(self.utils.val_X.float().cuda(device)).squeeze()
+                val_loss = mse(self.utils.val_Y.float().cuda(device), val_y_pred).detach().item()
+            self.net.train()
+            scheduler.step()
+            print("Epoch: " + str(epoch+1) + ", Loss: " + str(loss) + ", valLoss: " + str("%.5f" % val_loss) + ", Time: "+ str("%.1f" % (time.time() - startepoch) ) )
+        print("Loop time: ", time.time() - startloop)
+
+        # Save model for later use
+        torch.save({'epoch': epoch, 'model_state_dict': self.net.state_dict(), 'optimizer_state_dict': optimizer.state_dict(), 'scheduler_state_dict': scheduler.state_dict(),'loss': loss}, "post/finalepoch_tempsave.pth")
+
+    def inference(self, device):
+        self.net.to(device)
+        self.net.eval()
+        saved_dict = torch.load(self.manager.args["load_saved_model_from"])
+        self.net.load_state_dict(saved_dict['model_state_dict'])
+        scaler = joblib.load(self.manager.args["load_scaler_from"])
+        mse = nn.MSELoss()
+        # load testing set
+        utils = datautils(self.manager, self.opts, scaler)
+        utils.load_testing_set()
+
+        with torch.no_grad():
+            test_y_pred = self.net(utils.test_X.float().cuda(device)).squeeze()
+            test_loss = mse(utils.test_Y.float().cuda(device), test_y_pred).detach().item()
+        print("Testing Loss: " + str("%.5f" % test_loss))
+        nbins = 100
+        plt.hist(utils.test_Y.numpy(), bins=nbins, histtype = "step", color = "r", label = "Test Dataset")
+        plt.hist(test_y_pred.detach().cpu().numpy(), bins=nbins, histtype = "step", label = "DNN Prediction")
+        plt.xlabel("-log_10(DY weight)")
+        plt.ylabel("events")
+        plt.legend()
+        plt.savefig(self.manager.args["save_testing_histogram_at"])
```

Comment:
In another PR (no need to add more things for this one :+1:) as no interactive visualization is being done with `matplotlib` it would probably be better to use the OO API for this.

e.g.

```python
from matplotlib.figure import Figure

fig = Figure()
ax = fig.subplots()

ax.hist(...)
ax.set_xlabel(r"$-\log_{10}\,($Drell-Yan MoMEMta Weights$)$")
... 
```

For a bit of the motivations c.f. https://gist.github.com/matthewfeickert/84245837f09673b2e7afea929c016904 as this is related to `matplotlib` backends being set to work with or disallow GUIs.